db_stress_tool/expected_state.h

```diff
@@ -42,36 +214,49 @@ class ExpectedState {
   //    guaranteed finished. This is useful for crash-recovery testing when the
   //    process may crash before updating the expected values array.
   //
-  // Requires external locking covering `key` in `cf`.
-  void Put(int cf, int64_t key, uint32_t value_base, bool pending);
+  // Requires external locking covering `key` in `cf` to prevent concurrent
+  // write or delete to the same `key`.
+  ExpectedValue& Put(int cf, int64_t key, bool pending);
```

Comment:
Can it return an `ExpectedValue` (non-reference)?

Reply:
Actually I would rather eliminate `bool pending` argument and return a `PendingExpectedValue` that can be committed/aborted:

```
class PendingExpectedValue {
 public:
  explicit PendingExpectedValue(std::atomic<uint32_t>* value_ptr,
                                ExpectedValue orig_value,
                                ExpectedValue final_value)
      : value_ptr_(value_ptr),
        orig_value_(orig_value),
        final_value_(final_value) {}

  void Abort() {
    value_ptr_->store(orig_value_.Read(), std::memory_order_relaxed);
  }

  void Commit() {
    // prevent expected-value update from reordering before Write
    std::atomic_thread_fence(std::memory_order_release);
    value_ptr_->store(final_value_.Read(), std::memory_order_relaxed);
  }

  bool OriginalExists() { return orig_value_.Exists(); }

  uint32_t GetFinalValueBase() { return final_value_.GetValueBase(); }

 private:
  std::atomic<uint32_t>* const value_ptr_;
  const ExpectedValue orig_value_;
  const ExpectedValue final_value_;
};
```

The equivalent of `Put(..., true /* pending */)` would be `PreparePut()`:

```
void ExpectedState::Precommit(int cf, int64_t key, const ExpectedValue& value) {
  Value(cf, key).store(value.Read(), std::memory_order_relaxed);
  // prevent expected-value update from reordering after Write
  std::atomic_thread_fence(std::memory_order_release);
}

PendingExpectedValue ExpectedState::PreparePut(int cf, int64_t key) {
  ExpectedValue expected_value = Load(cf, key);
  const ExpectedValue orig_expected_value = expected_value;
  expected_value.Put(true /* pending */);
  const ExpectedValue pending_expected_value = expected_value;
  expected_value.Put(false /* pending */);
  const ExpectedValue final_expected_value = expected_value;
  Precommit(cf, key, pending_expected_value);
  return PendingExpectedValue(&Value(cf, key), orig_expected_value,
                              final_expected_value);
}
```

The equivalent of `Put(..., false /* pending */)` would be `PendingExpectedValue::Commit()` (shown above). Together, these could be used in `TestPut()`  like this:

```
    auto pending_expected_value =
        shared->PreparePut(rand_column_family, rand_key);
    const uint32_t value_base = pending_expected_value.GetFinalValueBase();
    ...
    DB::Put()
    ...
    pending_expected_value.Commit();
```

And similar for Delete(), SingleDelete(), etc

Reply:
Fixed

Reply:
The reason this became relevant in this PR is because this PR needed to introduce a coupling between `Put(cf, key, true /* pending */)` and `Put(cf, key, false /* pending */)`